delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/ResourceSourceParser.scala

```diff
@@ -28,39 +31,58 @@ trait ResourceSourceParser {
   )(implicit
       uuidF: UUIDF,
       rcr: RemoteContextResolution
-  ): IO[InvalidJsonLdFormat, (Iri, CompactedJsonLd, ExpandedJsonLd)] =
+  ): IO[InvalidJsonLdFormat, (Iri, CompactedJsonLd, ExpandedJsonLd)] = {
+    val (ctx, sourceWithCtx) = ctxOrProjectDefaults(source, project)
     for {
-      originalExpanded <- JsonLd.expand(source).leftMap(err => InvalidJsonLdFormat(None, err))
+      originalExpanded <- JsonLd.expand(sourceWithCtx).leftMap(err => InvalidJsonLdFormat(None, err))
       base              = project.base.iri
       iri              <- originalExpanded.rootId.asIri.fold(uuidF().map(uuid => base / uuid.toString))(IO.pure)
       expanded          = originalExpanded.replaceId(iri)
-      compacted        <- expanded.toCompacted(source).leftMap(err => InvalidJsonLdFormat(Some(iri), err))
+      compacted        <- expanded.toCompacted(ctx).leftMap(err => InvalidJsonLdFormat(Some(iri), err))
     } yield (iri, compacted, expanded)
+  }
 
   /**
     * Converts the passed ''source'' to JsonLD compacted and expanded.
     * The @id value is extracted from the payload if exists and compared to the passed ''iri''.
     * If they aren't equal an [[UnexpectedResourceId]] rejection is issued.
     *
+    * @param project the project used to generate the @context when no @context is provided on the source
     * @param source the Json payload
     * @return a tuple with the compacted Json-LD and the expanded Json-LD
     */
   def asJsonLd(
+      project: Project,
       iri: Iri,
       source: Json
-  )(implicit rcr: RemoteContextResolution): IO[ResourceRejection, (CompactedJsonLd, ExpandedJsonLd)] =
+  )(implicit rcr: RemoteContextResolution): IO[ResourceRejection, (CompactedJsonLd, ExpandedJsonLd)] = {
+    val (ctx, sourceWithCtx) = ctxOrProjectDefaults(source, project)
     for {
-      originalExpanded <- JsonLd.expand(source).leftMap(err => InvalidJsonLdFormat(Some(iri), err))
+      originalExpanded <- JsonLd.expand(sourceWithCtx).leftMap(err => InvalidJsonLdFormat(Some(iri), err))
       _                <- checkSameId(iri, originalExpanded)
       expanded          = originalExpanded.replaceId(iri)
-      compacted        <- expanded.toCompacted(source).leftMap(err => InvalidJsonLdFormat(Some(iri), err))
+      compacted        <- expanded.toCompacted(ctx).leftMap(err => InvalidJsonLdFormat(Some(iri), err))
     } yield (compacted, expanded)
+  }
 
   private def checkSameId(iri: Iri, expanded: ExpandedJsonLd): IO[UnexpectedResourceId, Unit] =
     expanded.rootId.asIri match {
       case Some(sourceId) if sourceId != iri => IO.raiseError(UnexpectedResourceId(iri, sourceId))
       case _                                 => IO.unit
     }
+
+  private def ctxOrProjectDefaults(source: Json, project: Project): (ContextValue, Json) =
+    source.topContextValueOrEmpty match {
```

Comment:
I'm wondering if we should check first to see if we have any triples... what if the json is fully expanded?

Reply:
What is the problem if the json is fully expanded? it will stay expanded as is

Reply:
We would unnecessarily add a context.